go/signatures_test.go

```diff
@@ -0,0 +1,336 @@
+}
+
+func sigTestCertLogEntry(t *testing.T) LogEntry {
+	return LogEntry{
+		Index: 0,
+		Leaf: MerkleTreeLeaf{
+			Version:  V1,
+			LeafType: TimestampedEntryLeafType,
+			TimestampedEntry: TimestampedEntry{
+				Timestamp: sigTestSCTTimestamp,
+				EntryType: X509LogEntryType,
+				X509Entry: mustDehex(t, sigTestDERCertString),
+			},
+		},
+	}
+}
+
+func mustCreateSignatureVerifier(t *testing.T, pk crypto.PublicKey) SignatureVerifier {
+	sv, err := NewSignatureVerifier(pk)
+	if err != nil {
+		t.Fatalf("Failed to create SignatureVerifier: %v", err)
+	}
+	return *sv
+}
+
+func TestVerifySCTSignatureEC(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTEC(t), sigTestCertLogEntry(t)); err != nil {
+		t.Fatalf("Failed to verify signature on SCT: %v", err)
+	}
+
+}
+
+func TestVerifySCTSignatureRSA(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestRSAPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTRSA(t), sigTestCertLogEntry(t)); err != nil {
+		t.Fatalf("Failed to verify signature on SCT: %v", err)
+	}
+
+}
+
+func TestVerifySCTSignatureFailsForMismatchedSignatureAlgorithm(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTRSA(t), sigTestCertLogEntry(t)); err == nil {
+		t.Fatalf("Sucessfully verified with mismatched signature algorithm")
+	}
+
+}
+
+func TestVerifySCTSignatureFailsForUnknownSignatureAlgorithm(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTWithSignature(t, sigTestCertSCTSignatureUnsupportedSignatureAlgorithm, sigTestKeyIDEC), sigTestCertLogEntry(t)); err == nil {
+		t.Fatal("Successfully verified signature with unsupported signature algorithm")
+	}
+}
+
+func TestVerifySCTSignatureFailsForUnknownHashAlgorithm(t *testing.T) {
+	v := mustCreateSignatureVerifier(t, sigTestECPublicKey(t))
+	if err := v.VerifySCTSignature(sigTestSCTWithSignature(t, sigTestCertSCTSignatureUnsupportedHashAlgorithm, sigTestKeyIDEC), sigTestCertLogEntry(t)); err == nil {
+		t.Fatal("Successfully verified signature with unsupported hash algorithm")
+	}
+}
+
+func testVerifySCTSignatureFailsForIncorrectInput(t *testing.T, sct SignedCertificateTimestamp, sv SignatureVerifier) {
+	entry := sigTestCertLogEntry(t)
+	for i := range entry.Leaf.TimestampedEntry.X509Entry {
+		old := entry.Leaf.TimestampedEntry.X509Entry[i]
+		entry.Leaf.TimestampedEntry.X509Entry[i] ^= byte(mrand.Intn(255) + 1)
+		if err := sv.VerifySCTSignature(sct, entry); err == nil {
+			t.Fatalf("Incorrectly verfied signature over corrupted input data, uncovered byte at %d?", i)
+		}
+		entry.Leaf.TimestampedEntry.X509Entry[i] = old
+	}
+}
+
+func testVerifySCTSignatureFailsForIncorrectSignature(t *testing.T, sct SignedCertificateTimestamp, sv SignatureVerifier) {
+	sct.Signature.Signature[mrand.Intn(len(sct.Signature.Signature))] ^= byte(mrand.Intn(255) + 1)
```

Comment:
Please send over a fixed byte index for the (probably) ASN.1 rendering of key material for all possible signature algorithms that we may end up supporting in the future which is guaranteed to be part of the key material and never a piece of metadata. :)

More seriously, doing it this way is going to cover things that occur in real life - flaky RAM, cosmic rays, dodgy routers, etc. don't tend to nicely flip just a single bit exactly where we want it to on demand.  So why *not* do it randomly? If this test ever passes then there's something interesting to look in to...